internal/controller/effective_dnspolicies_reconciler.go

```diff
@@ -50,6 +54,128 @@ func (r *EffectiveDNSPoliciesReconciler) Subscription() controller.Subscription
 	}
 }
 
+func isWildCardListener(l *machinery.Listener) bool {
+	return l != nil && (l.Hostname == nil || *l.Hostname == "" || strings.HasPrefix(string(*l.Hostname), "*."))
+}
+
+type hostRoutable struct {
+	hostnames []v1.Hostname
+}
+
+// DNSNamesForGateway returns a set of dnstargets to create records for keyed against a listener location
+func dnsNamesForGatewayFromRoutes(ctx context.Context, topology *machinery.Topology, gateway *machinery.Gateway) map[string][]dnsTarget {
+	logger := controller.LoggerFromContext(ctx).WithName("DNSNamesForGateway")
+	// this will give us all routes targeting the sectionName + any targeting the gateway directly
+	gatewayListeners := lo.FilterMap(topology.All().Children(gateway), func(item machinery.Object, _ int) (*machinery.Listener, bool) {
+		v, ok := item.(*machinery.Listener)
+		return v, ok
+	})
+	//TODO is there a better way to do this?
+	var gatewayHTTPRoutes = []*hostRoutable{}
+	for _, gl := range gatewayListeners {
+		// duplicated TODO clean up
+		hasAttachedRoute := false
+		for _, statusListener := range gateway.Status.Listeners {
+			if string(gl.Name) == string(statusListener.Name) {
+				hasAttachedRoute = statusListener.AttachedRoutes > 0
+				break
+			}
+		}
+		if !hasAttachedRoute {
+			continue
+		}
+
+		listenerRoutes := lo.FilterMap(topology.All().Children(gl), func(item machinery.Object, _ int) (*hostRoutable, bool) {
+			switch v := item.(type) {
+			case *machinery.GRPCRoute:
+				return &hostRoutable{hostnames: v.Spec.Hostnames}, true
+			case *machinery.HTTPRoute:
+				return &hostRoutable{hostnames: v.Spec.Hostnames}, true
+			default:
+				return nil, false
+			}
+		})
+		gatewayHTTPRoutes = append(gatewayHTTPRoutes, listenerRoutes...)
+	}
+
+	names := map[string][]dnsTarget{}
+	for _, route := range gatewayHTTPRoutes {
+		for _, routeHost := range route.hostnames {
+			var hostMatch, listernerID, listenerName string
+			for _, listener := range gatewayListeners {
+				if !isWildCardListener(listener) {
+					//exact match were done
+					if *listener.Hostname == routeHost {
+						logger.V(1).Info("exact listener host match found", "listener:", listener.GetLocator(), "route host", routeHost)
+						listernerID = listener.GetLocator()
+						listenerName = string(listener.Name)
+						break
+					}
+				}
+				if isWildCardListener(listener) {
```

Comment:
open question here should we remove this and allow *.a.b.com to be specified in a HTTPRoute and created. Seems against the concept